### PR TITLE
Add `.isWeakRef()`

### PR DESCRIPTION
**Pull request by bigbigDreamer**
@sindresorhus done other `PR` to replace #164

### CODE REVIEW

**Comment by sindresorhus** [comment]
Tests are failing

File: tsconfig.json [hunk 1]
```diff
@@ -2,9 +2,10 @@
 	"extends": "@sindresorhus/tsconfig",
 	"compilerOptions": {
 		"outDir": "dist",
-		"target": "es2018",
+		"target": "ES2018",
 		"lib": [
-			"es2018",
+		  	"ES2018",
```
Comment (sindresorhus) [inline tsconfig.json:7]:
```suggestion
			"ES2018",
```

File: package.json [hunk 2]
```diff
@@ -56,11 +56,11 @@
 		"ava": "^3.3.0",
 		"del-cli": "^2.0.0",
 		"eslint-config-xo-typescript": "^0.26.0",
-		"jsdom": "^16.0.1",
+		"jsdom": "^19.0.0",
 		"rxjs": "^6.4.0",
 		"tempy": "^0.4.0",
 		"ts-node": "^8.3.0",
-		"typescript": "~3.8.2",
+		"typescript": "^4.6.2",
```
Comment (sindresorhus) [inline package.json:63]:
```suggestion
		"typescript": "~4.6.2",
```

File: tsconfig.json (hunk 1 above)
Comment (bigbigDreamer) [reply tsconfig.json:7]: done

File: package.json (hunk 2 above)
Comment (bigbigDreamer) [reply package.json]: done

**Comment by bigbigDreamer** [comment]
> Tests are failing

Sorry, I haven't checked the problem yet. At present, it seems that only node version 14 is passed.

**Comment by bigbigDreamer** [comment]
> Tests are failing

image

I found the reason, because JSDOM WeakRef currently supports V14.6.0+ , can I Mock one manually? As a downgrade solution to handle Node versions？

**Comment by bigbigDreamer** [comment]
Unfortunately, I have not found a solution on how to simulate the behavior of `WeakRef`. At present, because `JSOM` simulation is limited to `Node V14.6.0+`, the test of the lower version fails, so I can only temporarily close this `PR`, thank you for your understanding

**Comment by bigbigDreamer** [comment]
https://github.com/ungap/weakrefs/blob/master/cjs/index.js  perhaps it can try mock in node env

**Comment by bigbigDreamer** [comment]
Although the use case must be run on a low version node environment, the minimum requirement for WeakRef is Node V14.6.0, so I don't think it is necessary to create a mock use case for it on a low version Node.

**Comment by bigbigDreamer** [comment]
Record:  JSDOM@latest will reported error in ci with Node V10.
image

**Comment by bigbigDreamer** [comment]
Tests all pass, @sindresorhus 🥳

**Comment by sindresorhus** [comment]
Can you fix the merge conflict?

**Comment by bigbigDreamer** [comment]
> Can you fix the merge conflict?

yeah. Had fixed！

image

File: source/index.ts [hunk 3]
```diff
@@ -608,8 +612,8 @@ export const assert: Assert = {
 	},
 	buffer: (value: unknown): asserts value is Buffer => assertType(is.buffer(value), 'Buffer', value),
 	blob: (value: unknown): asserts value is Blob => assertType(is.blob(value), 'Blob', value),
-	nullOrUndefined: (value: unknown): asserts value is null | undefined => assertType(is.nullOrUndefined(value), AssertionTypeDescription.nullOrUndefined, value), // eslint-disable-line @typescript-eslint/ban-types
-	object: (value: unknown): asserts value is object => assertType(is.object(value), 'Object', value), // eslint-disable-line @typescript-eslint/ban-types
+	nullOrUndefined: /* eslint-disable-line @typescript-eslint/ban-types */ (value: unknown): asserts value is null | undefined => assertType(is.nullOrUndefined(value), AssertionTypeDescription.nullOrUndefined, value),
+	object: (value: unknown): asserts value is Record<string, unknown> => assertType(is.object(value), 'Object', value),
```
Comment (sindresorhus) [inline source/index.ts:616]: Don't make unrelated changes. This change is incorrect.

**Comment by sindresorhus** [comment]
I suggest we just embed the `WeakRef` type so we don't have to require a higher target.

```ts
interface WeakRef<T extends object> {
    readonly [Symbol.toStringTag]: "WeakRef";

    /**
     * Returns the WeakRef instance's target object, or undefined if the target object has been
     * reclaimed.
     */
    deref(): T | undefined;
}
```

File: source/index.ts (hunk 3 above)
Comment (bigbigDreamer) [reply source/index.ts:616]: image
Because lint can't pass，I'm so sad.

**Comment by bigbigDreamer** [comment]
> I suggest we just embed the `WeakRef` type so we don't have to require a higher target.
> 
> ```ts
> interface WeakRef<T extends object> {
>     readonly [Symbol.toStringTag]: "WeakRef";
> 
>     /**
>      * Returns the WeakRef instance's target object, or undefined if the target object has been
>      * reclaimed.
>      */
>     deref(): T | undefined;
> }
> ```

done

File: source/index.ts (hunk 3 above)
Comment (bigbigDreamer) [reply source/index.ts]: > Don't make unrelated changes. This change is incorrect.

Ok, because I resolved a conflict and lost the lint disable rule, sorry

**Comment by bigbigDreamer** [comment]
> > Can you fix the merge conflict?
> 
> Ok. Had fixed！But I encountered a very strange problem. Line 343 is a blank line, but lint reported an error. I repeatedly checked the code and found no problems. Can I ask for your help.
> 
> image

Thank you so much, I just noticed that I read the wrong file, it is a lint problem in another file, sorry, sorry for the trouble, I will be more careful next time
